chore(core): add config hot-reload subsystem and implementation for pgwire credential changes

sklarsa opened the pull request:
This swaps out PropServerConfiguration usage with ReloadingPropServerConfiguration. It should be a no-op, in preparation for hot-reloading config

Instead of inheriting from PropServerConfiguration, I decided to wrap it in a new class to avoid moving code around too much, since the validation logic is in the PropServerConfiguration constructor

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 34 / 49 (69.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ReloadingPropServerConfiguration.java|33|48|68.75%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 34 / 49 (69.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ReloadingPropServerConfiguration.java|33|48|68.75%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 39 / 58 (67.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ReloadingPropServerConfiguration.java|34|53|64.15%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 38 / 57 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/ReloadingPropServerConfiguration.java|34|53|64.15%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|3|3|100.00%|

bluestreak01 commented on core/src/main/c/linux/filewatcher.c:
```diff
@@ -0,0 +1,179 @@
+    struct file_watcher *fw = (struct file_watcher *)wp;
+
+    // Clean up the inotify watch descriptor first
+    free(fw->wd);
```
this call looks extremely wonky, we need to free the following:
- all mallocs, which pointers have been assigned to the struct
- all fds that have been assigned

bluestreak01 commented on core/src/main/c/linux/filewatcher.c:
```diff
@@ -0,0 +1,179 @@
+    teardown(address);
+}
+
+int main(int argc, char *argv[])
```
this looks like leftover

bluestreak01 commented on core/src/main/c/linux/filewatcher.c:
```diff
@@ -0,0 +1,179 @@
+
+    /* Re-copy the path so we can extract the filename */
+    strcpy(pathCpy, fp);
+    char *filename = malloc(sizeof(char) * strlen(pathCpy) + 1);
```
this leaks

bluestreak01 commented on core/src/main/c/linux/filewatcher.c:
```diff
@@ -0,0 +1,179 @@
+    struct file_watcher *fw;
+
+    /* Create the file descriptor for accessing the inotify API. */
+    const int fd = inotify_init();
```
this leaks

bluestreak01 commented on core/src/main/java/io/questdb/FileWatcher.java:
```diff
@@ -0,0 +1,78 @@
+    @Override
+    public void close() {
+        if (!closed) {
+            reloadThread.interrupt();
```
interrupt cannot be used to stop the thread, it is just a flag that is being set. Nor interrupt waits for the thread to stop. There should be a different mechanism, that guarantees thread is halted and also gurantees lack of race conditions.

I am wondering if closing fd would release the wait? After that the thread should wind down gracefully?

sklarsa replied:
Yes, this is a recent addition that I wasn't sure about.  I'm removing the `interrupt` call, but still need to validate that all resources are properly released

bluestreak01 commented on core/src/main/c/linux/filewatcher.c:
```diff
@@ -0,0 +1,179 @@
+
+            if (event->len && !(event->mask & IN_ISDIR))
+            {
+                if (strcmp(event->name, fw->filename))
```
we are listening for single `fd`. What are the circumstances when we get notificed of something else?

bluestreak01 commented on core/src/main/c/linux/filewatcher.c:
```diff
@@ -0,0 +1,179 @@
+    }
+    fw->fd = fd;
+    fw->wd = wd;
+    fw->filename = filename;
```
i don't think we need to hold on to the filename

sklarsa replied:
I use `filename` to match filesystem events so we only emit them for that specific file: https://github.com/questdb/questdb/blob/0e9807f40c7f5b77a0646eb284d812c63a39b25e/core/src/main/c/linux/filewatcher.c#L137

bluestreak01 commented on core/src/main/c/linux/filewatcher.c:
```diff
@@ -0,0 +1,179 @@
+       inotify to watch the entire directory (and its contents) for changes
+       but only surface changes made to the watched file */
+    char pathCpy[strlen(fp)];
+    strcpy(pathCpy, fp);
```
we should not need a copy of string

sklarsa replied:
I found that when I didn't copy, the `dirname` call attempted to modify the `fp` string, which segfaulted because it is in protected memory.  Per man pages:

> Both dirname() and basename() may modify the contents of path, so it may be desirable to pass a copy when calling one of these functions. 
https://linux.die.net/man/3/basename

I did find an extra, unnecessary `strcpy` which I deleted in 1264dacad

bluestreak01 commented on core/src/main/java/io/questdb/InotifyFileEventNotifier.java:
```diff
@@ -0,0 +1,42 @@
+    public InotifyFileEventNotifier(CharSequence dirPath) {
+        try (Path p = new Path()) {
+            p.of(dirPath).$();
+            this.filewatcherPtr = setup(p.ptr());
```
the `setup` returns 0 in case of parameter error, we should at least have an assert here

bluestreak01 commented on core/src/main/java/io/questdb/FileWatcher.java:
```diff
@@ -0,0 +1,78 @@
+import io.questdb.std.Misc;
+import io.questdb.std.QuietCloseable;
+
+public class FileWatcher implements QuietCloseable {
```
we need a test around file watcher, to make sure it can produce callbacks and can halt without segmentation fault

bluestreak01 commented on core/src/main/java/io/questdb/InotifyFileEventNotifier.java:
```diff
@@ -0,0 +1,42 @@
+    private boolean closed;
+
+    public InotifyFileEventNotifier(CharSequence dirPath) {
+        try (Path p = new Path()) {
```
can be replaced with Path.getThreadLocal()

bluestreak01 reviewed (changes requested):
also formatting is out :)

sklarsa commented:
@bluestreak01 here's an update on this.

I just merged master into this branch, fixed merge conflicts, and all tests pass (with the exception of the flaky `TestPgWireCredentialsReloadWithNewProp`, related to this PR).

It looks like I was in the process of moving my `inotify` code into java to take advantage of the `epoll` helpers that we have. But then I moved on to other cloud-related things.  

The Bootstrap and ServerMain plumbing code is working as expected for the `DynamicServerConfiguration`. I also have the `DynamicUsernamePasswordMatcher` which takes advantage of the dynamic server configuration for hot-reloading PGWire credentials.

As for the cross-platform filewatcher stuff, here is where we are.

| OS | Method | Status |
| ---- | ----------- | --------- |
| Linux | Inotify | In progress |
| Osx | Kqueue | Needs review (and recompilation) |
| Freebsd | Kqueue | Needs review (and recompilation) |
| Windows | ? | Not started |

I'm happy to sync with you on this when you have time. I'd love to finally finish this thing.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/questdb/questdb?pullRequest=4168) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/questdb/questdb?pullRequest=4168) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: bluestreak01<br/>:white_check_mark: sklarsa<br/>:x: GitHub Actions - Rebuild Native Libraries<br/><hr/>**GitHub Actions - Rebuild Native Libraries** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/questdb/questdb?pullRequest=4168) it.</sub>

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 235 / 370 (63.51%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|70|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/FileWatcherNativeException.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/FileWatcherFactory.java|3|6|50.00%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/ServerMain.java|8|12|66.67%|
|:large_blue_circle:|io/questdb/InotifyFileWatcher.java|58|73|79.45%|
|:large_blue_circle:|io/questdb/InotifyAccessor.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/FileWatcher.java|22|26|84.62%|
|:large_blue_circle:|io/questdb/std/DebouncingRunnable.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 235 / 373 (63.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|73|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/FileWatcherNativeException.java|0|4|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/FileWatcherFactory.java|3|6|50.00%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/ServerMain.java|8|12|66.67%|
|:large_blue_circle:|io/questdb/InotifyFileWatcher.java|58|73|79.45%|
|:large_blue_circle:|io/questdb/InotifyAccessor.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/FileWatcher.java|22|26|84.62%|
|:large_blue_circle:|io/questdb/std/DebouncingRunnable.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 243 / 373 (65.15%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|73|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/FileWatcherFactory.java|3|6|50.00%|
|:large_blue_circle:|io/questdb/FileWatcherNativeException.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/ServerMain.java|8|12|66.67%|
|:large_blue_circle:|io/questdb/InotifyAccessor.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/InotifyFileWatcher.java|63|73|86.30%|
|:large_blue_circle:|io/questdb/std/DebouncingRunnable.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/FileWatcher.java|23|26|88.46%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 243 / 373 (65.15%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|73|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/FileWatcherFactory.java|3|6|50.00%|
|:large_blue_circle:|io/questdb/FileWatcherNativeException.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/ServerMain.java|8|12|66.67%|
|:large_blue_circle:|io/questdb/InotifyAccessor.java|4|5|80.00%|
|:large_blue_circle:|io/questdb/InotifyFileWatcher.java|63|73|86.30%|
|:large_blue_circle:|io/questdb/std/DebouncingRunnable.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/FileWatcher.java|23|26|88.46%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 254 / 378 (67.20%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|74|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|20|23|86.96%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/DebouncingRunnable.java|15|17|88.24%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|68|74|91.89%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 283 / 446 (63.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/DefaultFactoryProvider.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

bluestreak01 commented on core/src/main/c/freebsd/kqueue.c:
```diff
@@ -27,88 +27,168 @@
     return (jint) kevent(
-            kq,
-            (const struct kevent *) changelist,
+            kq, (const struct kevent *) changelist,
```
nit: formatting

bluestreak01 commented on core/src/main/c/freebsd/kqueue.c:
```diff
@@ -27,88 +27,168 @@
+JNIEXPORT jint JNICALL Java_io_questdb_KqueueAccessor_keventGetBlocking
+        (JNIEnv *e, jclass cl, jint kq, jlong eventList, jint nEvents) {
+    int res;
+    RESTARTABLE(kevent(kq, NULL, 0, (struct kevent *) eventList, nEvents, NULL), res);
```
what is happening with "res" here? it is passed by value and then returned, it is always 0

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 282 / 445 (63.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 282 / 445 (63.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 282 / 445 (63.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|78|86|90.70%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/Bootstrap.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 297 / 474 (62.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|96|88.54%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 297 / 474 (62.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|96|88.54%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 299 / 476 (62.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|96|88.54%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 299 / 476 (62.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|96|88.54%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 299 / 476 (62.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|96|88.54%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 299 / 476 (62.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|96|88.54%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

jerrinot commented on core/src/main/java/io/questdb/KqueueFileWatcher.java:
```diff
@@ -0,0 +1,219 @@
+                throw CairoException.critical(Os.errno()).put("could allocate kevent for directory");
+            }
+
+            long fds = KqueueAccessor.pipe();
```
should handle `pipe()` returning a negative value. when either native `pipe()` or `fcntl()` fails.

sklarsa replied:
```c
    res = fcntl(fds[0], F_SETFL, O_NONBLOCK);
    if (res < 0) {
        close(fds[0]);
        close(fds[1]);
        return res;
    }
    return (jlong) fds[0] << 32 | (jlong) fds[1];
```

Isn't it technically possible to return a valid negative value for res since the return value is a combination of 2 numbers and we're working with signed numbers?  If so, I can check `readEndFd` and `writeEndFd` separately for negative values after we perfom the bitshift...

jerrinot replied:
In theory - it's possible - when `int` has more than 32 bits. 

In practice, `int` has 32 bits on all supported systems. So both `FDs` have the highest bit set to `0`. (because valid FDs are positive). so `(jlong) fds[0] << 32 | (jlong) fds[1];` must have the highest bit set to `0` too -> the returned value cannot be negative.

jerrinot commented on core/src/main/java/io/questdb/KqueueFileWatcher.java:
```diff
@@ -0,0 +1,219 @@
+        }
+        // For now, we don't filter events down to the file,
+        // we trigger on every change in the directory
+        // todo: implement kqueue file filter
```
do we need the filter at all? perhaps this TODO can be just removed?

sklarsa replied:
nope.. not for the current use case

jerrinot commented on core/src/main/java/io/questdb/ServerConfiguration.java:
```diff
@@ -67,4 +67,9 @@ default String getReleaseType() {
     default void init(CairoEngine engine, FreeOnExit freeOnExit) {
     }
+
+    default boolean isLineTcpEnabled() {
```
why do we need this?

sklarsa replied:
no idea... I think it was an artifact from a merge

jerrinot commented:
@sklarsa: it's looking good, just some minor details left. The only bigger(?) remaining thing is EE counter-part. 

At very least we should ensure that:
1. EE compiles at all and all existing tests are passing
1. Whatever dynamic behaviour works in OSS also works with EE.

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 299 / 476 (62.82%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/ServerConfiguration.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|96|88.54%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

sklarsa commented:
> Whatever dynamic behaviour works in OSS also works with EE.

I need to wire up EE to get this working... I'll give it a shot in a PR over there

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 299 / 474 (63.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|95|89.47%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 299 / 474 (63.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|85|95|89.47%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 303 / 481 (62.99%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/Bootstrap.java|11|17|64.71%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|89|102|87.25%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

jerrinot commented:
a failure due to a flaky test: https://github.com/questdb/questdb/issues/4683

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 300 / 478 (62.76%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|89|102|87.25%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 302 / 480 (62.92%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|24|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|91|104|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 302 / 479 (63.05%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|23|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|91|104|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 302 / 479 (63.05%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|71|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|23|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|91|104|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

jerrinot commented:
unrelated failure in `testJsonQueryErrorOnDataUnavailableEventNeverFired()`, fixed by https://github.com/questdb/questdb/pull/4711

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 302 / 481 (62.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|73|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|23|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|91|104|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 302 / 481 (62.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/filewatch/WindowsAccessorImpl.java|0|15|00.00%|
|:large_blue_circle:|io/questdb/KqueueFileWatcher.java|0|73|00.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherWindows.java|0|23|00.00%|
|:large_blue_circle:|io/questdb/KqueueAccessor.java|0|21|00.00%|
|:large_blue_circle:|io/questdb/ServerMain.java|7|13|53.85%|
|:large_blue_circle:|io/questdb/Bootstrap.java|9|15|60.00%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcherFactory.java|5|8|62.50%|
|:large_blue_circle:|io/questdb/std/str/Utf8s.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/network/Epoll.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessor.java|5|6|83.33%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxAccessorFacadeImpl.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/DynamicPropServerConfiguration.java|91|104|87.50%|
|:large_blue_circle:|io/questdb/std/filewatch/FileWatcher.java|65|73|89.04%|
|:large_blue_circle:|io/questdb/std/filewatch/LinuxFileWatcher.java|67|73|91.78%|
|:large_blue_circle:|io/questdb/DynamicUsernamePasswordMatcher.java|28|29|96.55%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/PropBootstrapConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/DefaultPgWireAuthenticatorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/FactoryProviderImpl.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Os.java|1|1|100.00%|
